File: flutter/library/lib/src/tsp_calendar_heatmap.dart
import 'package:flutter/material.dart';
import 'star_planet_theme.dart';

class TspCalendarCell {
  const TspCalendarCell({required this.date, this.level = 'none'});

  final String date;
  final String level;
}

typedef TspCalendarDaySelected = void Function(String date, String level);

/// Month attendance heatmap.
class TspCalendarHeatmap extends StatelessWidget {
  const TspCalendarHeatmap({
    super.key,
    this.yearMonth,
    this.cells = const [],
    this.showLegend = true,
    this.theme = StarPlanetTheme.sky,
    this.onSelectDay,
  });

  final String? yearMonth;
  final List<TspCalendarCell> cells;
  final bool showLegend;
  final StarPlanetTheme theme;
  final TspCalendarDaySelected? onSelectDay;

  static const _weekdays = ['一', '二', '三', '四', '五', '六', '日'];
  static const _levels = [
    ('full', '全勤'),
    ('partial', '部分'),
    ('none', '未打'),
    ('exempt', '豁免'),
  ];

  static ({int year, int month}) _parseYearMonth(String? raw) {
    final parts = (raw ?? '').split('-');
    if (parts.length >= 2) {
      final y = int.tryParse(parts[0]);
      final m = int.tryParse(parts[1]);
      if (y != null && m != null && m >= 1 && m <= 12) return (year: y, month: m);
    }
    final now = DateTime.now();
    return (year: now.year, month: now.month);
  }

  ({Color bg, Color border, Color fg}) _colors(String level) {
    switch (level) {
      case 'full':
        return (bg: theme.success, border: theme.success, fg: Colors.white);
      case 'partial':
        return (bg: theme.warning, border: theme.warning, fg: theme.textPrimary);
      case 'exempt':
        return (bg: theme.brandPrimary, border: theme.brandPrimary, fg: Colors.white);
      default:
        return (bg: theme.pageEnd, border: theme.borderDefault, fg: theme.textTertiary);
    }
  }

  @override
  Widget build(BuildContext context) {
    final ym = _parseYearMonth(yearMonth);
    final levelMap = {for (final c in cells) c.date: c.level};
    final first = DateTime(ym.year, ym.month, 1);
    final daysInMonth = DateTime(ym.year, ym.month + 1, 0).day;
    final startPad = (first.weekday + 6) % 7; // Monday-first
    final slots = <({int day, String date, String level})?>[];
    for (var i = 0; i < startPad; i++) {
      slots.add(null);
    }
    for (var d = 1; d <= daysInMonth; d++) {
      final date =
          '${ym.year}-${ym.month.toString().padLeft(2, '0')}-${d.toString().padLeft(2, '0')}';
      slots.add((day: d, date: date, level: levelMap[date] ?? 'none'));
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        if (showLegend)
          Wrap(
            spacing: 6,
            runSpacing: 6,
            children: _levels.map((item) {
              final colors = _colors(item.$1);
              return Container(
                constraints: const BoxConstraints(minHeight: 28),
                padding: const EdgeInsets.symmetric(horizontal: 10),
                alignment: Alignment.center,
                decoration: BoxDecoration(
                  color: colors.bg,
                  borderRadius: BorderRadius.circular(999),
                  border: Border.all(color: colors.border),
                ),
                child: Text(
                  item.$2,
                  style: TextStyle(color: colors.fg, fontSize: 12, fontWeight: FontWeight.w800),
                ),
              );
            }).toList(),
          ),
        if (showLegend) const SizedBox(height: 10),
        Row(
          children: _weekdays
              .map(
                (w) => Expanded(
                  child: Text(
                    w,
                    textAlign: TextAlign.center,
                    style: TextStyle(color: theme.textTertiary, fontSize: 11, fontWeight: FontWeight.w800),
                  ),
                ),
              )
              .toList(),
        ),
        const SizedBox(height: 4),
        LayoutBuilder(
          builder: (context, constraints) {
            final gap = 4.0;
            final cellSize = (constraints.maxWidth - gap * 6) / 7;
            return Wrap(
              spacing: gap,
              runSpacing: gap,
              children: [
                for (final slot in slots)
                  SizedBox(
                    width: cellSize,
                    height: cellSize,
                    child: slot == null
                        ? const SizedBox.shrink()
                        : Builder(
                            builder: (_) {
                              final colors = _colors(slot.level);
                              return GestureDetector(
                                onTap: () => onSelectDay?.call(slot.date, slot.level),
                                child: DecoratedBox(
                                  decoration: BoxDecoration(
                                    color: colors.bg,
                                    borderRadius: BorderRadius.circular(8),
                                    border: Border.all(color: colors.border),
                                  ),
                                  child: Center(
                                    child: Text(
                                      '${slot.day}',
                                      style: TextStyle(
                                        color: colors.fg,
                                        fontSize: 13,
                                        fontWeight: FontWeight.w800,
                                      ),
                                    ),
                                  ),
                                ),
                              );
                            },
                          ),
                  ),
              ],
            );
          },
        ),
      ],
    );
  }
}

File: react-native/library/src/starPlanet/components/TspCalendarHeatmap.js
import React from 'react';
import { Pressable, StyleSheet, Text, View } from 'react-native';
import { withTheme } from '../utils/shared';

const WEEKDAYS = ['一', '二', '三', '四', '五', '六', '日'];
const LEVELS = [
  { level: 'full', label: '全勤' },
  { level: 'partial', label: '部分' },
  { level: 'none', label: '未打' },
  { level: 'exempt', label: '豁免' },
];

function parseYearMonth(yearMonth) {
  const [y, m] = String(yearMonth || '').split('-').map(Number);
  if (!y || !m) {
    const now = new Date();
    return { year: now.getFullYear(), month: now.getMonth() + 1 };
  }
  return { year: y, month: m };
}

function cellColors(level, t) {
  if (level === 'full') {
    return { backgroundColor: t.success, borderColor: t.success, color: '#FFFFFF' };
  }
  if (level === 'partial') {
    return { backgroundColor: t.warning, borderColor: t.warning, color: t.textPrimary };
  }
  if (level === 'exempt') {
    return { backgroundColor: t.brandPrimary, borderColor: t.brandPrimary, color: '#FFFFFF' };
  }
  return {
    backgroundColor: t.pageEnd,
    borderColor: t.borderDefault,
    color: t.textTertiary,
  };
}

/**
 * TspCalendarHeatmap – Month attendance heatmap.
 */
export function TspCalendarHeatmap({
  yearMonth,
  cells = [],
  showLegend = true,
  theme,
  onSelectDay,
}) {
  const t = withTheme(theme);
  const { year, month } = parseYearMonth(yearMonth);
  const levelMap = new Map(cells.map((c) => [c.date, c.level || 'none']));
  const first = new Date(year, month - 1, 1);
  const daysInMonth = new Date(year, month, 0).getDate();
  const startPad = (first.getDay() + 6) % 7;
  const slots = [];
  for (let i = 0; i < startPad; i += 1) slots.push(null);
  for (let d = 1; d <= daysInMonth; d += 1) {
    const date = `${year}-${String(month).padStart(2, '0')}-${String(d).padStart(2, '0')}`;
    slots.push({ day: d, date, level: levelMap.get(date) || 'none' });
  }

  return (
    <View style={styles.root}>
      {showLegend ? (
        <View style={styles.legend}>
          {LEVELS.map((item) => {
            const colors = cellColors(item.level, t);
            return (
              <View
                key={item.level}
                style={[
                  styles.legendItem,
                  {
                    backgroundColor: colors.backgroundColor,
                    borderColor: colors.borderColor,
                  },
                ]}
              >
                <Text style={[styles.legendText, { color: colors.color }]}>{item.label}</Text>
              </View>
            );
          })}
        </View>
      ) : null}
      <View style={styles.weekdays}>
        {WEEKDAYS.map((w) => (
          <Text key={w} style={[styles.weekday, { color: t.textTertiary }]}>
            {w}
          </Text>
        ))}
      </View>
      <View style={styles.grid}>
        {slots.map((slot, index) => {
          if (!slot) {
            return <View key={`empty-${index}`} style={[styles.cell, styles.cellEmpty]} />;
          }
          const colors = cellColors(slot.level, t);
          return (
            <Pressable
              key={slot.date}
              onPress={() => onSelectDay?.(slot.date, slot.level)}
              style={[
                styles.cell,
                {
                  backgroundColor: colors.backgroundColor,
                  borderColor: colors.borderColor,
                },
              ]}
            >
              <Text style={[styles.cellText, { color: colors.color }]}>{slot.day}</Text>
            </Pressable>
          );
        })}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  root: { gap: 10 },
  legend: { flexDirection: 'row', flexWrap: 'wrap', gap: 8 },
  legendItem: {
    borderWidth: 1,
    borderRadius: 999,
    paddingHorizontal: 10,
    paddingVertical: 4,
  },
  legendText: { fontSize: 11, fontWeight: '800' },
  weekdays: { flexDirection: 'row' },
  weekday: {
    width: `${100 / 7}%`,
    textAlign: 'center',
    fontSize: 12,
    fontWeight: '700',
  },
  grid: { flexDirection: 'row', flexWrap: 'wrap' },
  cell: {
    width: `${100 / 7}%`,
    aspectRatio: 1,
    borderWidth: 1,
    borderRadius: 10,
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 4,
  },
  cellEmpty: { borderWidth: 0, backgroundColor: 'transparent' },
  cellText: { fontSize: 13, fontWeight: '800' },
});
